### PR TITLE
fix: undo operation on signature pad

### PR DESCRIPTION
**Pull request by plxity**
fixes: #864

### CODE REVIEW

**Comment by plxity** [comment]
@Mythie @dguyen Please review the changes.

**Review by adithyaakrishna — APPROVED** [review]
LGTM 🌮

**Comment by plxity** [comment]
@Mythie Bumping this up 😅

**Comment by catalinpit** [comment]
I'm on the `main` branch, and the undo functionality works properly. Am I missing something?

https://github.com/documenso/documenso/assets/25515812/b1a94c93-39fa-452c-a08b-6fd7f590a088

**Comment by plxity** [comment]
@catalinpit Please watch the video I have attached and the steps to reproduce.

**Comment by catalinpit** [comment]
> @catalinpit Please watch the video I have attached and the steps to reproduce.

Yeah, I've seen those. I thought it happens in all scenarios, but the issue actually occurs when you already have a signature and try to update it.



Thanks for the heads up! Will test the changes and review the PR.

**Comment by catalinpit** [comment]
Ok, I just tested it. It works fine on the profile page, but for some reason doesn't work when signing a document. I attached a video.

Does it work for you?

https://github.com/documenso/documenso/assets/25515812/4bbb2142-e5f9-4696-bc8f-b4a575a94f72

**Comment by plxity** [comment]
Yup, it seems to be working for me. Can you restart server or clear browser cache and try again?

**Comment by catalinpit** [comment]
> Yup, it seems to be working for me. Can you restart server or clear browser cache and try again?

I restarted the server, cleared browser cache, and even ran `npm run reset:hard`, and it still doesn't work unfortunately...

**Comment by plxity** [comment]
@catalinpit My bad, I have fixed it. In `useEffect` deps value, there is no need to add `defaultValue`. Because we want that effect to only run once when the component loads, I've fixed it and have tested the changes at all the places. Please review it again.

File: packages/ui/primitives/signature-pad/signature-pad.tsx [hunk 1]
```diff
@@ -183,8 +184,11 @@ export const SignaturePad = ({
     // Clear the canvas
     if ($el.current) {
       const ctx = $el.current.getContext('2d');
+      const { width, height } = $el.current;
```
[the remark this answers is not in the document]
Comment (plxity) [reply packages/ui/primitives/signature-pad/signature-pad.tsx:187]: Fixed it, I think we can add eslint error for an unused variable😅

File: packages/ui/primitives/signature-pad/signature-pad.tsx [hunk 2]
```diff
@@ -183,8 +184,11 @@ export const SignaturePad = ({
     // Clear the canvas
     if ($el.current) {
```
Comment (catalinpit) [inline packages/ui/primitives/signature-pad/signature-pad.tsx:185]: Another thing I observed is that when you undo the first change to an already existing signature, the signature is not updated. See the video to better understand what I mean.

https://github.com/documenso/documenso/assets/25515812/281b776d-8366-4116-8bb9-0df5d6f1cfdb

I tried calling the `onChange` function after clearing the canvas to update the signature URL and it seems to work.

```js
if ($el.current) {
      const ctx = $el.current.getContext('2d');
      const { width, height } = $el.current;
      ctx?.clearRect(0, 0, width, height);
      if (typeof defaultValue === 'string' && defaultImageRef.current) {
        ctx?.putImageData(defaultImageRef.current, 0, 0);
      }
      newLines.forEach((line) => {
        const pathData = new Path2D(getSvgPathFromStroke(getStroke(line, perfectFreehandOptions)));
        ctx?.fill(pathData);
      });

      /*
         here
      */
      onChange?.($el.current.toDataURL());
    }
```

Please double check and see if it works and/or if there is a better way of doing it.

Here's how it works with the new addition.

https://github.com/documenso/documenso/assets/25515812/10a95739-e0b5-4889-894b-5d57df96c136

**Review by catalinpit — CHANGES_REQUESTED** [review]

File: packages/ui/primitives/signature-pad/signature-pad.tsx (hunk 2 above)
Comment (plxity) [reply packages/ui/primitives/signature-pad/signature-pad.tsx:184]: @catalinpit Thanks for pointing out. This makes sense, I've added an onChange handler. This is a good method to work with it :)

Comment (plxity) [reply packages/ui/primitives/signature-pad/signature-pad.tsx:184]: @catalinpit Let me know if the changes are fine :)

**Review by catalinpit — DISMISSED** [review]
thanks for the work!

**Review by Mythie — APPROVED** [review]

**Comment by Mythie** [comment]
Great work 😄

**Pull request merged**